### PR TITLE
setup e2e to run with aks-e clusters

### PR DESCRIPTION
**Pull request by aramase**
fixes https://github.com/Azure/secrets-store-csi-driver-provider-azure/issues/36

- [x] Update current pipeline to use go1.14
- [x] Add job for running against aks-engine cluster
  - [x] Setup pipeline to create and delete aks-engine cluster
  - [x] add make target for building and pushing to acr for testing
  - [x] run e2e tests on the aks-engine cluster with acr images

kind-e2e will be used as a gate to decide if tests need to be run with aks-engine cluster. If basic tests fail with kind-e2e, then test will terminate with error. For new tests that are added in `test/bats` that are azure specific, we can skip running test if `CI_KIND_CLUSTER` is true.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -45,30 +60,40 @@ unit-test:
...
 e2e-azure:
 	bats -t test/bats/azure.bats
 
-.PHONY: build-windows
-build-windows:
-	CGO_ENABLED=0 GOOS=windows go build -a -ldflags "-X main.BuildDate=$(BUILD_DATE) -X main.BuildVersion=$(IMAGE_VERSION)" -o _output/secrets-store-csi-driver-provider-azure.exe ./cmd/
+.PHONY: setup-kind
+setup-kind:
+	curl -L https://github.com/kubernetes-sigs/kind/releases/download/v${KIND_VERSION}/kind-linux-amd64 --output kind && chmod +x kind && sudo mv kind /usr/local/bin/
+	kind create cluster --config kind-config.yaml --image kindest/node:v${KIND_K8S_VERSION}
```
Comment (ritazh) [inline Makefile:95]: we don't need this config now that we are using k8s v1.16.3

File: .pipelines/e2e-job-azure.yaml [hunk 2]
```diff
@@ -0,0 +1,45 @@
+parameters:
+  name: ''
+
+jobs:
+- job: ${{ parameters.name }}
+  dependsOn:
+  - unit_tests
+  - e2e_test_kind
+  timeoutInMinutes: 120
```
Comment (ritazh) [inline .pipelines/e2e-job-azure.yaml:9]: We should probably timeout alot sooner than 120mins right?

File: .pipelines/kubernetes.json [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        "orchestratorProfile": {
+            "orchestratorType": "Kubernetes",
+            "orchestratorRelease": "$AKSENGINE_K8S_VERSION",
+            "kubernetesConfig": {
+                "loadBalancerSku": "Standard",
+                "excludeMasterFromStandardLB": true,
+                "addons": [
+                    {
+                        "name": "csi-secrets-store",
+                        "enabled": false
```
Comment (ritazh) [inline .pipelines/kubernetes.json:13]: If we are not using the addon in aks-engine, should we remove this whole section?

File: .pipelines/kubernetes.json [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+                    }
+                ]
+            }
+        },
+        "servicePrincipalProfile": {
+            "clientId": "$CLIENT_ID",
+            "secret": "$CLIENT_SECRET"
+        }
+    }
+}
```
Comment (ritazh) [inline .pipelines/kubernetes.json:46]: nit: newline

File: .pipelines/e2e-job-azure.yaml (hunk 2 above)
Comment (aramase) [reply .pipelines/e2e-job-azure.yaml:9]: Sure, we can bring it down to 30m given cluster creation time is < 15 mins on average.

File: .pipelines/templates/aks-engine-setup.yaml [hunk 5]
```diff
@@ -0,0 +1,70 @@
+steps:
+  - script: |
+      curl -LO https://github.com/Azure/aks-engine/releases/download/v${AKS_ENGINE_VERSION}/aks-engine-v${AKS_ENGINE_VERSION}-linux-amd64.tar.gz
+      tar -zxvf aks-engine-v${AKS_ENGINE_VERSION}-linux-amd64.tar.gz
```
Comment (ritazh) [inline .pipelines/templates/aks-engine-setup.yaml:4]: should we use https://aka.ms/aks-engine/aks-engine-k8s-e2e.tar.gz

File: .pipelines/templates/aks-engine-setup.yaml [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+      CLIENT_SECRET: $(AAD_CLIENT_SECRET)
+      SUBSCRIPTION_ID: $(SUBSCRIPTION_ID)
+
+  - script: |
+      az login -t microsoft.com --service-principal -u ${CLIENT_ID} -p ${CLIENT_SECRET}
+      az account set -s ${SUBSCRIPTION_ID}
+      VMSS_NAME=$(az vmss list -g ${AZURE_CLUSTER_NAME} --subscription ${SUBSCRIPTION_ID} --query "[].name" -o tsv)
+
+      # Assign user assigned identity to the VMSS
+      az vmss identity assign -g ${AZURE_CLUSTER_NAME} -n $VMSS_NAME --subscription ${SUBSCRIPTION_ID} --identities /subscriptions/${SUBSCRIPTION_ID}/resourcegroups/${RESOURCE_GROUP}/providers/Microsoft.ManagedIdentity/userAssignedIdentities/${USER_MSI_NAME} >> /dev/null
```
Comment (ritazh) [inline .pipelines/templates/aks-engine-setup.yaml:63]: This tests user assigned msi. what about system assigned?

File: .pipelines/kubernetes.json (hunk 3 above)
Comment (aramase) [reply .pipelines/kubernetes.json:13]: `csi-secrets-store` addon is enabled by default for 1.16+ so, we are explicitly disabling the `csi-secrets-store` addon so we can install it as part of the e2e tests.

File: .pipelines/templates/aks-engine-setup.yaml (hunk 5 above)
Comment (aramase) [reply .pipelines/templates/aks-engine-setup.yaml:4]: Does it always host the latest version of aks-engine? Currently I've configured the pipeline to use `v0.49.0.`

File: .pipelines/templates/aks-engine-setup.yaml (hunk 6 above)
Comment (aramase) [reply .pipelines/templates/aks-engine-setup.yaml:63]: We can't perform role assignment for the system assigned MSI as we are logged in using SP. For user assigned MSI, we are using an existing one for which role assignment has already been done against existing keyvault.

File: Makefile [hunk 7]
```diff
@@ -2,29 +2,44 @@ IMAGE_NAME=secrets-store-csi-driver-provider-azure
 REGISTRY_NAME ?= upstreamk8sci
 REGISTRY ?= $(REGISTRY_NAME).azurecr.io
 DOCKER_IMAGE ?= $(REGISTRY)/public/k8s/csi/secrets-store/provider-azure
-IMAGE_VERSION ?= 0.0.4
+IMAGE_VERSION ?= 0.0.5
```
Comment (ritazh) [inline Makefile:5]: Should we update this when we are ready to make a new release?

File: .pipelines/templates/teardown.yaml [hunk 8]
```diff
@@ -0,0 +1,11 @@
+steps:
+  - script: |
+      rm ${LOCAL_SSH_KEY}*
+      az login -t microsoft.com --service-principal -u ${CLIENT_ID} -p ${CLIENT_SECRET}
+      az group delete -n ${AZURE_CLUSTER_NAME} --subscription ${SUBSCRIPTION_ID} --yes --no-wait
```
Comment (ritazh) [inline .pipelines/templates/teardown.yaml:5]: Should we also clean up the e2e image after the job is done?

Comment (aramase) [reply .pipelines/templates/teardown.yaml:5]: Done!

File: .pipelines/templates/aks-engine-setup.yaml (hunk 5 above)
Comment (ritazh) [reply .pipelines/templates/aks-engine-setup.yaml:4]: Right now it's pointing to `v0.48.0` but let's confirm with @chewong to see how often that gets updated to the latest.

Comment (ritazh) [reply .pipelines/templates/aks-engine-setup.yaml:4]: We can merge as is and replace it later.

File: .pipelines/templates/aks-engine-setup.yaml (hunk 6 above)
Comment (ritazh) [reply .pipelines/templates/aks-engine-setup.yaml:63]: ack.

**Pull request merged**